contracts/staking/stakeManager/StakeManager.sol

```diff
@@ -479,79 +495,199 @@ contract StakeManager is IStakeManager, StakeManagerStorage {
         reward = Math.min(CHECKPOINT_REWARD, reward);
 
         uint256 _proposerBonus = reward.mul(proposerBonus).div(100);
-        Validator storage _proposer = validators[signerToValidator[proposer]];
-        if (_proposer.contractAddress != address(0x0)) {
-            IValidatorShare(_proposer.contractAddress).addProposerBonus(_proposerBonus, _proposer.amount);
+        uint256 proposerId = signerToValidator[proposer];
+        Validator storage _proposer = validators[proposerId];
+        uint256 delegatedAmount = _proposer.delegatedAmount;
+        if (delegatedAmount > 0) {
+            _increaseValidatorRewardWithDelegation(
+                proposerId, 
+                _proposer.amount, 
+                delegatedAmount, 
+                _proposerBonus
+            );
         } else {
             _proposer.reward = _proposer.reward.add(_proposerBonus);
         }
 
-        reward = reward.sub(_proposerBonus);
-        uint256 stakePower = currentValidatorSetTotalStake();
         // update stateMerkleTree root for accounts balance on heimdall chain
         accountStateRoot = stateRoot;
-        _finalizeCommit();
-        return checkSignature(stakePower, reward, voteHash, sigs);
+        checkSignature(currentValidatorSetTotalStake(), reward.sub(_proposerBonus), voteHash, sigs);
```

Comment:
@imyourm8 Move 490 to 513 in `_increaseRewardAndAssertConsensus`, previously it was design bottleneck due to rewards distribution now we can do it after.
- it will save gas in terms of failed tx.
- accurate rewards calculation with accurate 2/3 power.

Reply:
I don't get it, 490 is a comment line

Reply:
You mean assert first before `rewardPerStake ` modification?

Reply:
I am talking about moving whole block 490-513.

Reply:
got it